interfaces/builtin/udisks2.go

```diff
@@ -431,24 +428,39 @@ func (iface *udisks2Interface) AppArmorConnectedPlug(spec *apparmor.Specificatio
 }
 
 func (iface *udisks2Interface) AppArmorPermanentSlot(spec *apparmor.Specification, slot *snap.SlotInfo) error {
-	if !release.OnClassic {
+	if !implicitSystemPermanentSlot(slot) {
 		spec.AddSnippet(udisks2PermanentSlotAppArmor)
 	}
 	return nil
 }
 
 func (iface *udisks2Interface) UDevPermanentSlot(spec *udev.Specification, slot *snap.SlotInfo) error {
-	if !release.OnClassic {
-		spec.AddSnippet(udisks2PermanentSlotUDev)
-		spec.TagDevice(`SUBSYSTEM=="block"`)
-		// # This tags all USB devices, so we'll use AppArmor to mediate specific access (eg, /dev/sd* and /dev/mmcblk*)
-		spec.TagDevice(`SUBSYSTEM=="usb"`)
+	if !implicitSystemPermanentSlot(slot) {
+		var udevFile string
+		slot.Attr("udev-file", &udevFile)
+		if udevFile != "" {
+			mountDir := slot.Snap.MountDir()
+			resolvedPath, err := osutil.ResolvePathNoEscape(mountDir, udevFile)
+			if err != nil {
+				return err
```

Comment:
we need a test for this case here now? also we probably need to wrap this error to give a bit more context?

Reply:
I added tests and wrapped it. Also did it for the reading of the file.